Remote Keyboard added

derFogel opened the pull request:
I added Remote input with a virtual keyboard with different layouts.
Currently there are only two layouts, but it is easy to add new ones (simply place a json file in keyboard-layouts/).

I tried to copy your coding style but I am not sure if i matched your expectations.

lievenhey committed:
added keyboard

keyboard added, currently only demo

derFogel committed:
added basic input features

modifiers (shift, ctrl, alt) not working

lievenhey committed:
keyboard now fully working, layout still missing

add option to change layout

lievenhey committed:
added keyboard change functionality

added english keyboard

lievenhey committed:
added translation

lievenhey committed:
restore old pro

lievenhey committed:
added missing modifiers resets

lievenhey committed:
added feedback and key peating

R1tschY commented:
Thank you for your pull request.

Why do you not use the system keyboard? For example through an invisible text control. This would remove the burden for creating and maintaining keyboard layouts and all the logic behind it. IMHO it is not necessary to create an own keyboard implementation. When it is possible, please use the system keyboard.

derFogel commented:
I made a own keyboard because the default keyboard is missing some keys (like ctrl, alt, arrows).
But yes, it is easier to use the default keyboard.
I will remove my own keyboard and use the system keyboard. Maybe it is possible to add these missing keys as an extra row to the default keyboard. Since it is based on the maliit keyboard, there should be something like a plugin api. I will talk with on of the developers of the maliit keyboard and see if something like this is possible.

R1tschY commented:
I did not thought about the extra keys. Maybe a hybrid approach makes sense: A keyboard with ASCII chars and control keys (ctrl, alt, win, arrow keys, ...) like the Terminal app keyboard (for shortcuts?) and the possibility to open the system keyboard to write texts in the user language.

R1tschY commented on app/src/ui/keyboardlayoutprovider.cpp:
```diff
@@ -117,3 +119,55 @@ QVariantList KeyboardLayoutProvider::row6() const
+
+void KeyboardLayoutProvider::saveConfig() const
+{
+    QFile config(QStandardPaths::writableLocation(QStandardPaths::ConfigLocation) + "/harbour-sailfishconnect/keyboardlayout.conf");
```
Please use QSettings for application-wide settings. Use "RemoteKeyboard" as group. As example see the runInBackground setting in the UI class: 
https://github.com/R1tschY/harbour-sailfishconnect/blob/2d6145c5f084370251d66b5b6a6e8f6cf1223528/app/src/ui.cpp#L49
https://github.com/R1tschY/harbour-sailfishconnect/blob/2d6145c5f084370251d66b5b6a6e8f6cf1223528/app/src/ui.cpp#L138

lievenhey committed:
improved keyboard layouts

layouts are now loaded from the default layouts in
/usr/share/maliit/plugins/com/jolla/layouts/

lievenhey committed:
renamed char key to ?123 and ABC

lievenhey committed:
Added QSettings groups

removed one files from SailfishConnect.pro (used it for testing, forgit
it)
removed unused function declarations

derFogel commented:
I modified the keyboard layout provider to use the default layout from jolla (they are stored in /usr/share/maliit/plugins/com/jolla/layouts/). It automatically adds the control and alt key next to the spacebar and adds the arrow keys when switching to the character page. I think this is a good solution because we don't have to maintain different keyboard layouts and every user can use their favourite layout.
I also changed my settings implementation (forgot about QSettings) to QSettings.

R1tschY commented:
Is your pull request ready for review?

derFogel commented:
yes

R1tschY commented:
@derFogel: Could you please resolve the conflicts in app.pro.

R1tschY commented on app/src/ui/keyboardlayoutprovider.cpp:
```diff
@@ -0,0 +1,223 @@
+    // seems to be the only way to get the files stored
+    QFile layoutFile("/usr/share/maliit/plugins/com/jolla/layouts/" + layout + ".qml");
+    if (!layoutFile.open(QIODevice::ReadOnly)) {
+        qDebug() << "unkown layout: " << layout;
```
Typo and please use an own logger for KeyboardLayoutProvider (see other files in app/src/ui).
```suggestion
        qCDebug(logger) << "Unknown layout:" << layout;
```

R1tschY commented on app/qml/components/KeyboardPage.qml:
```diff
@@ -0,0 +1,247 @@
+                            fillMode: Image.PreserveAspectFit
+                            visible: (modelData["symView"] === "left" || modelData["symView"] === "down" || modelData["symView"] === "right") && page.chars
+
+                            source:  "image://theme/icon-m-" + modelData["symView"]
```
I'm getting:
`[W] unknown:197 - file:///usr/share/harbour-sailfishconnect/qml/components/KeyboardPage.qml:197:25: QML Image: Failed to get image from provider: image://theme/icon-m-undefined`

R1tschY commented on app/qml/components/KeyboardChangeLayoutPage.qml:
```diff
@@ -0,0 +1,85 @@
+                    }
+                }
+
+                TextSwitch {
```
[W] unknown:74 - file:///usr/share/harbour-sailfishconnect/qml/components/KeyboardChangeLayoutPage.qml:74:17: QML TextSwitch: Binding loop detected for property "checked"

R1tschY commented on app/qml/components/KeyboardChangeLayoutPage.qml:
```diff
@@ -0,0 +1,85 @@
+
+            footer: Column {
+
+                Slider {
```
[W] unknown:61 - file:///usr/share/harbour-sailfishconnect/qml/components/KeyboardChangeLayoutPage.qml:61:17: QML Slider: Binding loop detected for property "value"

R1tschY commented on app/qml/components/KeyboardChangeLayoutPage.qml:
```diff
@@ -0,0 +1,85 @@
+    SilicaFlickable {
+        anchors.fill: parent
+
+        ListView {
```
Please use SilicaListView
```suggestion
        SilicaListView {
```

R1tschY commented on app/qml/components/KeyboardChangeLayoutPage.qml:
```diff
@@ -0,0 +1,85 @@
+Page {
+    allowedOrientations: Orientation.All
+
+    SilicaFlickable {
```
This SilicaFlickable is unneeded. Place everything into the SilicaListView.

R1tschY commented on app/qml/components/KeyboardChangeLayoutPage.qml:
```diff
@@ -0,0 +1,85 @@
+                }
+            }
+
+            footer: Column {
```
Please put it in the header. So it is a little bit hidden at the end of a long list.

R1tschY commented on app/qml/components/KeyboardChangeLayoutPage.qml:
```diff
@@ -0,0 +1,85 @@
+import QtQuick 2.0
+import Sailfish.Silica 1.0
+
+Page {
```
I think putting the layout list into a ComboBox provides a better workflow. What do you think?

derFogel replied:
It is definitely better. Thanks for the suggestion

R1tschY commented on app/qml/components/KeyboardPage.qml:
```diff
@@ -0,0 +1,247 @@
+        }
+    }
+
+    // I hate this hack but it seems this is the only way to push an attachted page
```
I would prefer the use of the pulley menu.

R1tschY commented on app/src/plugins/remoteinput/remoteinput.cpp:
```diff
@@ -0,0 +1,73 @@
+    //{Qt::Key_F12, 32},
+};
+
+SailfishConnect::RemoteInputPlugin::RemoteInputPlugin(Device *device, const QString &name, const QSet<QString> &outgoingCapabilities) : KdeConnectPlugin(device, name, outgoingCapabilities)
```
Please do not create lines with more than 80 charaters in C++ files. You can configure this limit in QtCreator. Please look through all your code to fix this.

R1tschY commented on app/src/plugins/remoteinput/remoteinput.cpp:
```diff
@@ -0,0 +1,73 @@
+#include "remoteinput.h"
```
Missing copyright header. Please look through all your code to fix this.

R1tschY reviewed (changes requested):
I did a short review of your changes.
Generally: Feedback and Emoij layout did not work for me.

lievenhey committed:
applied codereview improvements

lievenhey committed:
added combobox

derFogel committed:
Merge branch 'master' into master

lievenhey committed:
split up too long lines

lievenhey committed:
Merge branch 'master' of github.com:derFogel/harbour-sailfishconnect

derFogel commented:
@R1tschY done

R1tschY commented on app/qml/components/KeyboardPage.qml:
```diff
@@ -53,6 +53,15 @@ Page {
 
+        PullDownMenu {
+            MenuItem {
+                text: qsTr("Change layout")
```
Rename it to "Settings" or make two entries: "Change layout" and "Settings"

R1tschY commented on app/src/plugins/remoteinput/remoteinput.cpp:
```diff
@@ -0,0 +1,95 @@
+{
+}
+
+bool SailfishConnect::RemoteInputPlugin::receivePacket(const NetworkPacket &np)
```
Please handle "kdeconnect.mousepad.echo" and "kdeconnect.mousepad.keyboardstate" like in:
https://github.com/KDE/kdeconnect-kde/blob/7df58d0d04b5e7a1f99c3d35a8fd485bc83e868c/plugins/remotekeyboard/remotekeyboardplugin.cpp#L79

R1tschY commented on app/src/plugins/remoteinput/metadata.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "Id": "SailfishConnect::RemoteInputPlugin",
```
Please, rename every thing to remotekeyboard / remotekeyboardplugin / RemoteKeyboardPlugin. So the names match with KDE-Connect.

R1tschY commented on app/qml/components/KeyboardPage.qml:
```diff
@@ -0,0 +1,243 @@
+    function displayCorrectChar(data) {
+        if (page.chars && typeof data["symView"] !== "undefined") {
+            // dont show text on arrow keys
+            if (data["symView"] === "up" || data["symView"] === "left" || data["symView"] === "down" || data["symView"] === "right") return " ";
```
Please use something like ['a', 'b', 'c'].indexOf(str) >= 0 or str.match("a|b|c") to reduce line length

R1tschY commented on app/qml/components/KeyboardPage.qml:
```diff
@@ -0,0 +1,243 @@
+                                modifiers.alt = 0;
+                                modifiers.ctrl = 0;
+                            } else if (label == "alt") {
+                                modifiers.alt++;
```
Just `modifiers.alt = (modifiers.alt + 1) % 3`

R1tschY commented on app/qml/components/KeyboardPage.qml:
```diff
@@ -0,0 +1,243 @@
+
+                        onClicked: {
+                            if (modelData["caption"] === "shift") {
+                                modifiers.shift++;
```
Just `modifiers.shift = (modifiers.shift + 1) % 3`

R1tschY commented on app/qml/components/KeyboardPage.qml:
```diff
@@ -0,0 +1,243 @@
+                            width: parent.width
+                            height: parent.height
+                            fillMode: Image.PreserveAspectFit
+                            visible: (modelData["caption"] === "shift" && !page.chars) || modelData["caption"] === "backspace" || (modelData["symView"] === "up" && page.chars)
```
Please break lines in qml files for more readability (<= 100 good, <= 80 better).

R1tschY commented on app/src/ui/keyboardlayoutprovider.cpp:
```diff
@@ -0,0 +1,240 @@
+    QFile layoutFile("/usr/share/maliit/plugins/com/jolla/layouts/" + layout
+                     + ".qml");
+    if (!layoutFile.open(QIODevice::ReadOnly)) {
+        qCDebug(logger) << "unkown layout: " << layout;
```
```suggestion
        qCDebug(logger) << "Unknown layout: " << layout;
```

R1tschY commented on app/src/ui.cpp:
```diff
@@ -79,6 +80,8 @@ void UI::showMainWindow()
     // view
     m_view->rootContext()->setContextProperty("daemon", m_daemon);
     m_view->rootContext()->setContextProperty("ui", this);
+    m_view->rootContext()->setContextProperty("_keyboardLayout", m_keyboardLayoutProvider);
```
Why use an underscore? 
```suggestion
    m_view->rootContext()->setContextProperty("keyboardLayout", m_keyboardLayoutProvider);
```

R1tschY commented on app/qml/components/KeyboardPage.qml:
```diff
@@ -0,0 +1,243 @@
+                                if (modifiers.alt > 2) modifiers.alt = 0;
+                            } else if (label == "ctrl") {
+                                modifiers.ctrl++;
+                                if (modifiers.ctrl > 2) modifiers.ctrl = 0;
```
`modifiers.alt = (modifiers.alt + 1) % 3`

R1tschY commented on app/qml/components/KeyboardPage.qml:
```diff
@@ -0,0 +1,243 @@
+                                modifiers.ctrl = 0;
+                            } else if (label == "alt") {
+                                modifiers.alt++;
+                                if (modifiers.alt > 2) modifiers.alt = 0;
```
`modifiers.alt = (modifiers.alt + 1) % 3`

R1tschY commented on app/src/plugins/remoteinput/remoteinput.cpp:
```diff
@@ -0,0 +1,95 @@
+    bool shift, bool ctrl, bool alt) const
+{
+    NetworkPacket np("kdeconnect.mousepad.request", {{"key", key},
+                        {"specialKey", specialKeysMap[key]}, {"shift", shift},
```
Avoid creating new keys in specialKeysMap:
```suggestion
                        {"specialKey", specialKeysMap.value(key, 0)}, {"shift", shift},
```

R1tschY commented on app/src/ui.cpp:
```diff
@@ -79,6 +80,8 @@ void UI::showMainWindow()
     m_view->rootContext()->setContextProperty("daemon", m_daemon);
     m_view->rootContext()->setContextProperty("ui", this);
+    m_view->rootContext()->setContextProperty("_keyboardLayout", m_keyboardLayoutProvider);
+    qmlRegisterType<KeyboardLayoutProvider>("com.meego.maliitquick", 1, 0, "Layout");
```
Please do not reuse others namespaces. Please move the code to registerQmlTypes() in sailfishconnect.cpp.
```suggestion
    qmlRegisterType<KeyboardLayoutProvider>("SailfishConnect.RemoteKeyboard", 0, 4, "KeyboardLayout");
```

R1tschY commented on app/src/ui/keyboardlayoutprovider.cpp:
```diff
@@ -0,0 +1,240 @@
+/*
```
Please move it to the plugins/remotekeyboard dir.

R1tschY commented on app/app.pro:
```diff
@@ -80,7 +82,12 @@ DISTFILES += \
+    qml/components/Key.qml \
+    qml/components/RemoteInput.qml \
+    src/plugins/remoteinput/metadata.json \
+    keyboard-layouts/english.json
```
This file does not exist.

R1tschY reviewed (changes requested):
I played with it and missed the ESC key. Also there are many other keys missing (https://github.com/KDE/kdeconnect-kde/blob/7df58d0d04b5e7a1f99c3d35a8fd485bc83e868c/plugins/remotekeyboard/remotekeyboardplugin.cpp#L32). Any possibility to add a new row at the top?

At the moment landscape orientation is forced. I think it is a little bit strict. Why not allow other orientations?

Keep going. Looks promising. :)

R1tschY committed:
Update app/src/ui/keyboardlayoutprovider.cpp

Co-Authored-By: derFogel <derFogel@users.noreply.github.com>

lievenhey committed:
applied suggested changes

lievenhey committed:
Merge branch 'master' of github.com:derFogel/harbour-sailfishconnect

derFogel commented:
I am not sure how to handle the "kdeconnect.mousepad.echo" and "kdeconnect.mousepad.keyboardstate". The official andoid app also does not handled them (see https://github.com/KDE/kdeconnect-android/blob/master/src/org/kde/kdeconnect/Plugins/MousePadPlugin/KeyListenerView.java#L99).

R1tschY commented:
> I am not sure how to handle the "kdeconnect.mousepad.echo" and "kdeconnect.mousepad.keyboardstate". The official andoid app also does not handled them (see https://github.com/KDE/kdeconnect-android/blob/master/src/org/kde/kdeconnect/Plugins/MousePadPlugin/KeyListenerView.java#L99).

Just return true, so there is no warning about a not processed message.

lievenhey committed:
removed unprocessed message warning

derFogel commented:
ok, done

R1tschY commented:
Next week I will do a final review.

R1tschY commented on app/src/plugins/remotekeyboard/keyboardlayoutprovider.cpp:
```diff
@@ -0,0 +1,259 @@
+
+QVariantList KeyboardLayoutProvider::layouts()
+{
+    QDir layoutDir("/usr/share/maliit/plugins/com/jolla/layouts/");
```
When parsing "/usr/share/maliit/plugins/com/jolla/layouts*.conf" (QSettings can be used for parsing) you get all layouts with the correct language name.

R1tschY commented on app/qml/components/KeyboardChangeLayoutPage.qml:
```diff
@@ -0,0 +1,78 @@
+            anchors.fill: parent
+
+            PageHeader {
+                title: qsTr("Choose layout")
```
```suggestion
                title: qsTr("Settings")
```

R1tschY commented on app/qml/components/KeyboardPage.qml:
```diff
@@ -0,0 +1,278 @@
+        }
+    }
+
+    PageHeader {
```
Please use as header of SilicaFlickable:
```
SilicaFlickable {
  header: PageHeader { ...
```

derFogel replied:
Does not work: Error: Error while loading page: file:///usr/share/harbour-sailfishconnect/qml/components/KeyboardPage.qml:52 Cannot assign to non-existent property "header"

R1tschY reviewed (changes requested):
I found out that the language layout selection page is blank, when opening in landscape. But is not blank, when changing to landscape when page is already open. Would you fix this?

Using the layouts_*.conf files is not important. You can do it if you want.

R1tschY committed:
Update app/qml/components/KeyboardChangeLayoutPage.qml

Co-Authored-By: derFogel <derFogel@users.noreply.github.com>

Commit:
added long names

derFogel commented:
Done, now the displayed names are loaded from the config files

Commit:
reenabled tests
